src/main/java/org/broadinstitute/hellbender/tools/spark/BaseRecalibratorSparkOptimized.java

```diff
@@ -0,0 +1,205 @@
+            read -> BAQ.getReferenceWindowForRead(read, BAQ.DEFAULT_BANDWIDTH, BAQ.DEFAULT_INCLUDE_CLIPPED_BASES);
+
+
+    @Override
+    protected void runPipeline( JavaSparkContext ctx ) {
+        if ( readArguments.getReadFilesNames().size() != 1 ) {
+            throw new UserException("Sorry, we only support a single reads input for now.");
+        }
+        final String bam = readArguments.getReadFilesNames().get(0);
+        final String referenceURL = referenceArguments.getReferenceFileName();
+
+        auth = getAuthHolder();
+
+        final ReferenceMultiSource rds = new ReferenceMultiSource(auth, referenceURL, BQSR_REFERENCE_WINDOW_FUNCTION);
+
+        final ReadsHeaderSource readsDataflowSource = new ReadsHeaderSource(bam, auth);
+        SAMFileHeader readsHeader = readsDataflowSource.getHeader();
+        final SAMSequenceDictionary readsDictionary = readsHeader.getSequenceDictionary();
+        final SAMSequenceDictionary refDictionary = rds.getReferenceSequenceDictionary(readsDictionary);
+        final CountingReadFilter readFilterToApply = BaseRecalibrator.getStandardBQSRReadFilter(readsHeader);
+
+        Broadcast<SAMFileHeader> readsHeaderView = ctx.broadcast(readsHeader);
+        Broadcast<SAMSequenceDictionary> refDictionaryView = ctx.broadcast(refDictionary);
+
+        SequenceDictionaryUtils.validateDictionaries("reference", refDictionary, "reads", readsDictionary);
+
+        List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(readsHeader.getSequenceDictionary())
+                : IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
+
+        Stopwatch hacking = Stopwatch.createUnstarted();
+        List<String> localVariants = knownVariants;
+        if (BucketUtils.isCloudStorageUrl(knownVariants.get(0))) {
+            // TODO: fix
+            logger.info("(HACK): copying the GCS file to local just so we can read it back.");
+            hacking.start();
+            localVariants = new ArrayList<>();
+            // this only works with the API_KEY, but then again it's a hack so there's no point in polishing it.
+            PipelineOptions popts = auth.asPipelineOptionsOldStyle();
+            int i=0;
+            for (String s : knownVariants) {
+                // presumably they're all GCS.
+                String d = IOUtils.createTempFile("knownVariants-"+i,".vcf").getAbsolutePath();
+                try {
+                    BucketUtils.copyFile(s, popts, d);
+                } catch (IOException x) {
+                    throw new UserException.CouldNotReadInputFile(s,x);
+                }
+                i++;
+                localVariants.add(d);
+            }
+            hacking.stop();
+            logger.info("Copying the vcf took "+hacking.elapsed(TimeUnit.MILLISECONDS)+" ms.");
+        }
```

Comment:
Yup, done that already. But please add support for reading variant files from GCS so we can remove this hack.